File: client/src/ts/icepush/net/events/NewPlayerEvent.ts
import {NetworkEvent} from "../NetworkEvent";
import {Player} from "../../entity/Player";
import {PositionedBuffer} from "../../../engine/net/PositionedBuffer";
import {OPCode} from "../NetworkEventBuffer";

/**
 * This event is sent from the server to the client indicating a new player has been added to the game.
 */
export class NewPlayerEvent extends NetworkEvent {

    private readonly BINARY_SIZE: number;

    public readonly playerID: number;
    public readonly type: Player.Type;
    public readonly username: string;
    public readonly lives: number;

    /**
     * Creates an event from the given player and playerID.
     *
     * @param playerID The player's ID.
     * @param player The player object.
     */
    constructor(playerID: number, player: Player);

    /**
     * Reads the new player's data from the buffer.
     * @param buffer The buffer to read from.
     */
    constructor(buffer: PositionedBuffer);

    /**
     * Overload constructor.
     */
    constructor(bufferOrID: PositionedBuffer|number, player?: Player) {
        super();
        if (bufferOrID instanceof PositionedBuffer) {
            this.playerID = bufferOrID.readInt16BE();
            this.type = bufferOrID.readInt8();
            this.username = bufferOrID.readString();
            this.lives = bufferOrID.readUInt8();
        } else if (player !== undefined) {
            this.playerID = bufferOrID;
            this.type = player.getType();
            this.username = player.getName();
            this.lives = player.getLives();
        } else {
            throw new Error(`Malformed constructor:\n${bufferOrID}\n${player}`);
        }
        this.BINARY_SIZE = 4 + PositionedBuffer.getStringWriteSize(this.username);
    }

    /**
     * @override
     */
    public getEventSize(): number {
        return this.BINARY_SIZE;
    }

    /**
     * @override
     */
    public getOPCode(): OPCode {
        return OPCode.NEW_PLAYER;
    }

    /**
     * @override
     */
    public write(buffer: PositionedBuffer): void {
        buffer.writeUInt16BE(this.playerID);
        buffer.writeUInt8(this.type);
        buffer.writeString(this.username);
        buffer.writeUInt8(this.lives);
    }

}

File: src/net/threesided/server/net/event/events/server/NewPlayerEvent.java
package net.threesided.server.net.event.events.server;

import net.threesided.server.Player;
import net.threesided.server.net.event.OPCode;
import net.threesided.server.net.event.ServerNetworkEvent;
import net.threesided.shared.PacketBuffer;

import java.util.Collection;

public class NewPlayerEvent extends ServerNetworkEvent {

    private final short id;
    private final byte type, lives;
    private final String username;

    /**
     * Creates a new event to notify all players of a new player logging into the game.
     * @param player The new player.
     */
    public NewPlayerEvent(final Collection<PacketBuffer> recipients, final Player player) {
        this(recipients, player.getID(), player.getType().getID(), player.getUsername(), player.getLives());
    }

    /**
     * Creates a new event to notify all players of a new player logging into the game.
     * @param id The player's ID.
     * @param type The player's type.
     * @param username The player's username.
     * @param lives The number of lives the player has.
     */
    public NewPlayerEvent(final Collection<PacketBuffer> recipients,
                          final short id, final byte type,
                          final String username, final byte lives) {
        super(recipients);
        this.id = id;
        this.type = type;
        this.username = username;
        this.lives = lives;
    }

    @Override
    protected void writeDataToBuffer(final PacketBuffer buffer) {
        buffer.writeShort(this.id);
        buffer.writeByte(this.type);
        buffer.writeString(this.username);
        buffer.writeByte(this.lives);
    }

    @Override
    public OPCode getOPCode() {
        return OPCode.NEW_PLAYER;
    }

}
